lib/pmdtester/parsers/pmd_report_document.rb

```diff
@@ -8,80 +8,84 @@ class PmdReportDocument < Nokogiri::XML::SAX::Document
     attr_reader :violations
     attr_reader :errors
     attr_reader :configerrors
+    attr_reader :infos_by_rules
+
     def initialize(branch_name, working_dir, filter_set = nil)
-      @violations = PmdViolations.new
-      @errors = PmdErrors.new
-      @configerrors = PmdConfigErrors.new
+      @violations = CollectionByFile.new
+      @errors = CollectionByFile.new
+      @configerrors = {}
+
+      @infos_by_rules = {}
       @current_violations = []
       @current_violation = nil
       @current_error = nil
       @current_configerror = nil
-      @current_element = ''
-      @filename = ''
       @filter_set = filter_set
       @working_dir = working_dir
       @branch_name = branch_name
+
+      @cur_text = String.new(capacity: 200)
     end
 
     def start_element(name, attrs = [])
       attrs = attrs.to_h
-      @current_element = name
 
       case name
       when 'file'
-        remove_work_dir!(attrs['name'])
+        @current_filename = remove_work_dir!(attrs['name'])
         @current_violations = []
-        @current_filename = attrs['name'].freeze
       when 'violation'
         @current_violation = PmdViolation.new(attrs, @branch_name, @current_filename)
       when 'error'
-        remove_work_dir!(attrs['filename'])
         remove_work_dir!(attrs['msg'])
-        @current_filename = attrs['filename']
-        @current_error = PmdError.new(attrs, @branch_name)
-      when 'configerror'
-        @current_configerror = PmdConfigError.new(attrs, @branch_name)
+        @current_filename = remove_work_dir!(attrs['filename'])
+        @current_error = PmdError.new(attrs, @branch_name, @current_filename)
       end
     end
 
     def remove_work_dir!(str)
       str.sub!(/#{@working_dir}/, '')
+      str
     end
 
     def characters(string)
-      @current_violation&.text += string
+      @cur_text << remove_work_dir!(string)
```

Comment:
I think it's not really needed but it also doesn't hurt. If at some point the xml renderer decides to put exception messages outside of a cdata when it can, it should not break this class